File: backend/tour-design-api/src/main/java/com/tourdesign/platform/service/ClientPreferenceService.java
package com.tourdesign.platform.service;

import com.tourdesign.platform.entity.ClientPreferenceEntity;
import com.tourdesign.platform.model.ClientPreferenceModel;
import java.util.List;
import java.util.Optional;

public interface ClientPreferenceService {

    List<ClientPreferenceModel> list();

    ClientPreferenceModel create(ClientPreferenceModel obj);

    List<ClientPreferenceModel> getPreferencesByClient(Long clientId);

    Optional<ClientPreferenceModel> search(Long id);

    List<ClientPreferenceModel> searchAllById(List<Long> preferenceIds);

    Optional<ClientPreferenceModel> update(Long id, ClientPreferenceModel obj);

    boolean delete(Long id);

    List<ClientPreferenceModel> searchByClient(Long clientId);
}

File: frontend/tour-design-web/src/app/services/client-preference.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { ClientPreferenceModel } from '../models/client-preference.model';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ClientPreferenceService {
  private baseUrl = '/client-preference';

  constructor(private api: ApiService) {}

  list(): Observable<ClientPreferenceModel[]> {
    return this.api.get<ClientPreferenceModel[]>(this.baseUrl);
  }

  search(id: number): Observable<ClientPreferenceModel> {
    return this.api.get<ClientPreferenceModel>(`${this.baseUrl}/${id}`);
  }

  create(obj: ClientPreferenceModel): Observable<ClientPreferenceModel> {
    return this.api.post<ClientPreferenceModel>(this.baseUrl, obj);
  }

  update(id: number, obj: ClientPreferenceModel): Observable<ClientPreferenceModel> {
    return this.api.put<ClientPreferenceModel>(`${this.baseUrl}/${id}`, obj);
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchByClient(clientId: number): Observable<ClientPreferenceModel[]> {
    return this.api.get<ClientPreferenceModel[]>(`${this.baseUrl}/searchByClient/${clientId}`);
  }

}